File: 21-merge-sorted-list/go/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

const MaxUint = ^uint(0)
const MaxInt = int(MaxUint >> 1)

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	d := &ListNode{Val: 0, Next: nil}
	i := d

	for l1 != nil || l2 != nil {
		a := MaxInt
		b := MaxInt

		if l1 != nil {
			a = l1.Val
		}

		if l2 != nil {
			b = l2.Val
		}

		if a < b {
			i.Next = l1
			l1 = l1.Next
		} else {
			i.Next = l2
			l2 = l2.Next
		}

		i = i.Next
	}

	return d.Next
}

File: 21-merge-sorted-list/cpp/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
 public:
  ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
    ListNode* d = new ListNode();
    ListNode* i = d;

    while (l1 != NULL || l2 != NULL) {
      int v1 = l1 == NULL ? INT_MAX : l1->val;
      int v2 = l2 == NULL ? INT_MAX : l2->val;

      if (v1 < v2) {
        i->next = l1;
        l1 = l1->next;
      } else {
        i->next = l2;
        l2 = l2->next;
      }
      i = i->next;
    }
    return d->next;
  }
};
